File: js/test/regression-84.test.js
#!/usr/bin/env bun
/**
 * Regression tests for issue #84: "We should not run bash inside bash"
 *
 * These tests guard against the shell-inside-shell regression where
 * `$ --isolated docker -- bash` caused:
 *   docker run ... image /bin/bash -i -c bash   (WRONG: bash inside bash)
 * instead of:
 *   docker run ... image bash -i                 (CORRECT: bare shell with explicit -i)
 *
 * The -i flag is required so that bash reliably starts in interactive mode,
 * since docker's PTY bridging through Node.js spawn may not set up a TTY that
 * bash auto-detects. Without -i, bash may exit immediately (post-fix regression).
 *
 * The same regression applies to zsh, sh, and all other isolation backends.
 *
 * Also tests the post-fix regression hint: when a bare shell exits with code 1
 * quickly (< 3s), a helpful hint suggests `bash --norc` as a workaround for
 * containers whose .bashrc causes bash to exit non-zero (issue #84, second comment).
 *
 * Reference: https://github.com/link-foundation/start/issues/84
 * Fixed in: PR #85 (v0.24.1) via isInteractiveShellCommand()
 * Fixed in: PR #87 adding explicit -i for bare shells in docker attached mode
 */

const { describe, it } = require('node:test');
const assert = require('assert');
const { isInteractiveShellCommand } = require('../src/lib/isolation');
const { isDockerAvailable } = require('../src/lib/docker-utils');

// Helper: mirrors the command-args construction logic used in
// runInDocker attached mode.
// If this helper returns args containing '-c' for a bare shell command,
// the shell-inside-shell bug is present.
function buildCmdArgs(command, shellToUse = '/bin/bash') {
  const path = require('path');
  const shellName = shellToUse.split('/').pop();
  const shellInteractiveFlag =
    shellName === 'bash' || shellName === 'zsh' ? '-i' : null;
  const shellArgs = shellInteractiveFlag
    ? [shellToUse, shellInteractiveFlag]
    : [shellToUse];
  if (isInteractiveShellCommand(command)) {
    // Bare shell: pass directly with explicit -i for bash/zsh (issue #84 fix)
    const parts = command.trim().split(/\s+/);
    const bareShellFlag =
      path.basename(parts[0]) === 'bash' || path.basename(parts[0]) === 'zsh'
        ? '-i'
        : null;
    if (bareShellFlag && !parts.includes(bareShellFlag)) {
      return [parts[0], bareShellFlag, ...parts.slice(1)];
    }
    return parts;
  }
  return [...shellArgs, '-c', command];
}

describe('isInteractiveShellCommand additional cases (issue #84)', () => {
  // These cover edge cases not in the base isInteractiveShellCommand test suite.

  // Workaround: bash --norc skips .bashrc sourcing (post-fix regression workaround)
  it('should return true for "bash --norc"', () => {
    assert.strictEqual(isInteractiveShellCommand('bash --norc'), true);
  });

  it('should return true for "zsh --no-rcs"', () => {
    assert.strictEqual(isInteractiveShellCommand('zsh --no-rcs'), true);
  });

  it('should return true for "bash -i" (interactive flag, no -c)', () => {
    assert.strictEqual(isInteractiveShellCommand('bash -i'), true);
  });

  it('should return true for "fish"', () => {
    assert.strictEqual(isInteractiveShellCommand('fish'), true);
  });

  it('should return true for "dash"', () => {
    assert.strictEqual(isInteractiveShellCommand('dash'), true);
  });

  it('should return true for "/usr/local/bin/bash"', () => {
    assert.strictEqual(isInteractiveShellCommand('/usr/local/bin/bash'), true);
  });

  it('should return false for \'bash -c "echo hello"\'', () => {
    assert.strictEqual(
      isInteractiveShellCommand('bash -c "echo hello"'),
      false
    );
  });
});

describe('Regression: No Shell-Inside-Shell (issue #84)', () => {
  // Each test verifies that the command-arg construction logic does NOT
  // wrap a bare shell invocation inside another shell with `-c`.
  //
  // Before fix (v0.24.0): buildCmdArgs('bash') → ['/bin/bash', '-i', '-c', 'bash']
  // After fix (v0.24.1):  buildCmdArgs('bash') → ['bash', '-i']
  //   (bare shell passed directly with explicit -i for reliable interactive mode)

  it('should pass "bash" with -i flag, not wrap in shell -c', () => {
    const args = buildCmdArgs('bash');
    assert.deepStrictEqual(
      args,
      ['bash', '-i'],
      `Expected ["bash", "-i"], got: ${JSON.stringify(args)}`
    );
    assert.ok(
      !args.includes('-c'),
      'Must not contain -c flag (shell-inside-shell)'
    );
  });

  it('should pass "zsh" with -i flag, not wrap in shell -c', () => {
    const args = buildCmdArgs('zsh');
    assert.deepStrictEqual(args, ['zsh', '-i']);
    assert.ok(
      !args.includes('-c'),
      'Must not contain -c flag (shell-inside-shell)'
    );
  });

  it('should pass "sh" directly without -i (sh does not use -i flag), not wrap in shell -c', () => {
    const args = buildCmdArgs('sh', 'sh');
    assert.deepStrictEqual(args, ['sh']);
    assert.ok(
      !args.includes('-c'),
      'Must not contain -c flag (shell-inside-shell)'
    );
  });

  it('should pass "/bin/bash" with -i flag, not wrap in shell -c', () => {
    const args = buildCmdArgs('/bin/bash');
    assert.deepStrictEqual(args, ['/bin/bash', '-i']);
    assert.ok(
      !args.includes('-c'),
      'Must not contain -c flag (shell-inside-shell)'
    );
  });

  it('should pass "bash --norc" with -i flag (workaround for broken .bashrc)', () => {
    const args = buildCmdArgs('bash --norc');
    assert.deepStrictEqual(args, ['bash', '-i', '--norc']);
    assert.ok(!args.includes('-c'), 'Must not contain -c flag');
  });

  it('should pass "bash -l" directly (login shell, already has its own init)', () => {
    const args = buildCmdArgs('bash -l');
    assert.deepStrictEqual(args, ['bash', '-i', '-l']);
    assert.ok(!args.includes('-c'), 'Must not contain -c flag');
  });

  it('should still wrap non-shell commands in shell -c (guard against over-broad fix)', () => {
    const args = buildCmdArgs('echo hello', '/bin/bash');
    assert.deepStrictEqual(args, ['/bin/bash', '-i', '-c', 'echo hello']);
    assert.ok(
      args.includes('-c'),
      'Non-shell commands must still use -c wrapper'
    );
  });

  it('should still wrap "npm test" in shell -c (guard against over-broad fix)', () => {
    const args = buildCmdArgs('npm test', '/bin/bash');
    assert.deepStrictEqual(args, ['/bin/bash', '-i', '-c', 'npm test']);
    assert.ok(
      args.includes('-c'),
      'Non-shell commands must still use -c wrapper'
    );
  });

  it('should not treat "bash -c something" as bare shell', () => {
    // bash -c ... has -c, so isInteractiveShellCommand returns false
    // The command gets wrapped: ['/bin/bash', '-i', '-c', 'bash -c "echo hi"']
    const args = buildCmdArgs('bash -c "echo hi"', '/bin/bash');
    assert.ok(
      args.includes('-c'),
      'bash -c commands should be treated as regular commands'
    );
  });

  it('should not duplicate -i if user already passes "bash -i"', () => {
    // If user explicitly passes -i, we should not add another -i
    const args = buildCmdArgs('bash -i');
    assert.deepStrictEqual(
      args,
      ['bash', '-i'],
      'Should not duplicate -i flag'
    );
    assert.ok(!args.includes('-c'), 'Must not contain -c flag');
  });
});

describe('Docker daemon availability check (issue #84)', () => {
  // Verifies that isDockerAvailable returns a boolean (not a crash),
  // and that the error message for a non-running daemon is helpful.

  it('isDockerAvailable should return a boolean', () => {
    const result = isDockerAvailable();
    assert.strictEqual(
      typeof result,
      'boolean',
      'isDockerAvailable() must return a boolean'
    );
  });

  it('runInDocker error message for non-running daemon should be actionable', () => {
    // Mirrors the message in runInDocker when isDockerAvailable() returns false.
    const message =
      'Docker is installed but not running. Please start Docker Desktop or the Docker daemon, then try again.';
    assert.ok(
      message.includes('not running'),
      'Message must indicate Docker is not running'
    );
    assert.ok(
      message.includes('Docker Desktop'),
      'Message must mention Docker Desktop (common on macOS/Windows)'
    );
    assert.ok(
      message.includes('try again'),
      'Message must tell user what to do next'
    );
  });
});

describe('Post-fix regression hint: --norc suggestion (issue #84)', () => {
  // These tests verify the hint logic that recommends --norc when a bare shell
  // exits quickly with code 1 (e.g., broken .bashrc in konard/sandbox image).
  // The hint is shown in runInDocker attached mode when:
  //   isBareShell && code !== 0 && durationMs < 3000

  // Helper mirrors the hint construction logic in isolation.js runInDocker.
  const path = require('path');
  function buildHint(command) {
    const shellName = command.trim().split(/\s+/)[0];
    const noRcFlag = path.basename(shellName) === 'zsh' ? '--no-rcs' : '--norc';
    return (
      `Hint: The shell exited immediately — its startup file (.bashrc/.zshrc) may have errors.\n` +
      `Try skipping startup files: ${shellName} ${noRcFlag}`
    );
  }

  it('should suggest --norc for bare bash that exits quickly', () => {
    const hint = buildHint('bash');
    assert.ok(hint.includes('--norc'), 'Hint must include --norc for bash');
    assert.ok(
      hint.includes('bash --norc'),
      'Hint must show the full corrected command'
    );
    assert.ok(
      hint.includes('startup file'),
      'Hint must explain why the shell exited'
    );
  });

  it('should suggest --no-rcs for bare zsh that exits quickly', () => {
    const hint = buildHint('zsh');
    assert.ok(hint.includes('--no-rcs'), 'Hint must include --no-rcs for zsh');
    assert.ok(
      hint.includes('zsh --no-rcs'),
      'Hint must show the full corrected command'
    );
  });

  it('should suggest --norc for /bin/bash path that exits quickly', () => {
    const hint = buildHint('/bin/bash');
    assert.ok(
      hint.includes('/bin/bash --norc'),
      'Hint must include full path with --norc'
    );
  });

  it('should suggest --norc for bare sh that exits quickly', () => {
    const hint = buildHint('sh');
    assert.ok(hint.includes('sh --norc'), 'Hint must suggest --norc for sh');
  });

  it('should confirm workaround: bash --norc is detected as bare shell', () => {
    // The workaround (bash --norc) must still be recognized as a bare shell
    // so it gets passed directly to docker without -c wrapping
    assert.strictEqual(
      isInteractiveShellCommand('bash --norc'),
      true,
      'bash --norc must be a bare shell (no -c wrapping)'
    );
    assert.strictEqual(
      isInteractiveShellCommand('zsh --no-rcs'),
      true,
      'zsh --no-rcs must be a bare shell (no -c wrapping)'
    );
  });
});

File: rust/tests/regression_84_test.rs
//! Regression tests for issue #84: "We should not run bash inside bash"
//!
//! These tests guard against the shell-inside-shell regression where
//! `$ --isolated docker -- bash` caused:
//!   docker run ... image /bin/bash -i -c bash   (WRONG: bash inside bash)
//! instead of:
//!   docker run ... image bash -i                 (CORRECT: bare shell with explicit -i)
//!
//! Reference: https://github.com/link-foundation/start/issues/84
//! Fixed in: PR #85 (v0.24.1) via is_interactive_shell_command()

use start_command::{is_interactive_shell_command, is_shell_invocation_with_args};

mod is_interactive_shell_command_tests {
    use super::*;

    #[test]
    fn should_return_true_for_bash() {
        assert!(is_interactive_shell_command("bash"));
    }

    #[test]
    fn should_return_true_for_zsh() {
        assert!(is_interactive_shell_command("zsh"));
    }

    #[test]
    fn should_return_true_for_sh() {
        assert!(is_interactive_shell_command("sh"));
    }

    #[test]
    fn should_return_true_for_bash_norc() {
        assert!(is_interactive_shell_command("bash --norc"));
    }

    #[test]
    fn should_return_true_for_zsh_no_rcs() {
        assert!(is_interactive_shell_command("zsh --no-rcs"));
    }

    #[test]
    fn should_return_true_for_bash_i_interactive_flag() {
        assert!(is_interactive_shell_command("bash -i"));
    }

    #[test]
    fn should_return_true_for_fish() {
        assert!(is_interactive_shell_command("fish"));
    }

    #[test]
    fn should_return_true_for_dash() {
        assert!(is_interactive_shell_command("dash"));
    }

    #[test]
    fn should_return_true_for_full_path_bash() {
        assert!(is_interactive_shell_command("/usr/local/bin/bash"));
    }

    #[test]
    fn should_return_false_for_bash_c_echo_hello() {
        assert!(!is_interactive_shell_command("bash -c \"echo hello\""));
    }

    #[test]
    fn should_return_false_for_npm_test() {
        assert!(!is_interactive_shell_command("npm test"));
    }

    #[test]
    fn should_return_false_for_echo_hello() {
        assert!(!is_interactive_shell_command("echo hello"));
    }

    #[test]
    fn should_return_false_for_empty_string() {
        assert!(!is_interactive_shell_command(""));
    }
}

mod build_cmd_args_regression_84_tests {
    use super::*;

    /// Helper that mirrors the attached-mode command-args construction logic in run_in_docker.
    /// Returns the argv array that would be passed to `docker run ... image <argv>`.
    fn build_cmd_args(command: &str, shell_to_use: &str) -> Vec<String> {
        let shell_name = shell_to_use.rsplit('/').next().unwrap_or(shell_to_use);
        let shell_interactive_flag = match shell_name {
            "bash" | "zsh" => Some("-i"),
            _ => None,
        };

        if is_interactive_shell_command(command) {
            // Bare shell: pass directly with explicit -i flag (issue #84 fix)
            let parts: Vec<&str> = command.split_whitespace().collect();
            let basename = parts[0].rsplit('/').next().unwrap_or(parts[0]);
            let bare_flag = match basename {
                "bash" | "zsh" => Some("-i"),
                _ => None,
            };
            if let Some(flag) = bare_flag {
                if !parts.contains(&flag) {
                    let mut result = vec![parts[0].to_string(), flag.to_string()];
                    result.extend(parts[1..].iter().map(|s| s.to_string()));
                    return result;
                }
            }
            return parts.iter().map(|s| s.to_string()).collect();
        } else if is_shell_invocation_with_args(command) {
            // Shell with -c: pass directly as argv (issue #91 fix)
            return start_command::build_shell_with_args_cmd_args(command);
        }

        // Regular command: wrap in shell -c
        let mut shell_cmd_args = vec![shell_to_use.to_string()];
        if let Some(flag) = shell_interactive_flag {
            shell_cmd_args.push(flag.to_string());
        }
        shell_cmd_args.push("-c".to_string());
        shell_cmd_args.push(command.to_string());
        shell_cmd_args
    }

    #[test]
    fn bash_should_get_i_flag_not_wrapped_in_shell_c() {
        let args = build_cmd_args("bash", "/bin/bash");
        assert_eq!(args, vec!["bash", "-i"]);
        assert!(
            !args.contains(&"-c".to_string()),
            "Must not contain -c flag"
        );
    }

    #[test]
    fn zsh_should_get_i_flag_not_wrapped_in_shell_c() {
        let args = build_cmd_args("zsh", "/bin/bash");
        assert_eq!(args, vec!["zsh", "-i"]);
        assert!(
            !args.contains(&"-c".to_string()),
            "Must not contain -c flag"
        );
    }

    #[test]
    fn sh_should_not_get_i_flag_not_wrapped_in_shell_c() {
        let args = build_cmd_args("sh", "/bin/bash");
        assert_eq!(args, vec!["sh"]);
        assert!(
            !args.contains(&"-c".to_string()),
            "Must not contain -c flag"
        );
    }

    #[test]
    fn npm_test_should_be_wrapped_in_shell_c() {
        let args = build_cmd_args("npm test", "/bin/bash");
        assert!(
            args.contains(&"-c".to_string()),
            "npm test should be wrapped in shell -c"
        );
        assert!(args.contains(&"npm test".to_string()));
    }

    #[test]
    fn echo_hello_should_be_wrapped_in_shell_c() {
        let args = build_cmd_args("echo hello", "/bin/bash");
        assert!(
            args.contains(&"-c".to_string()),
            "echo hello should be wrapped in shell -c"
        );
    }

    #[test]
    fn bash_already_has_i_flag_no_duplicate() {
        let args = build_cmd_args("bash -i", "/bin/bash");
        let i_count = args.iter().filter(|a| a.as_str() == "-i").count();
        assert_eq!(i_count, 1, "Should not have duplicate -i flags: {:?}", args);
    }
}
